File: lua/ht/core/right-click/component/menu-item.lua
local MenuText = require("ht.core.right-click.component.text")
local UtilsTable = require("ht.utils.table")

---@class ht.right_click.MenuItemOptions
---@field text ht.right_click.MenuText
---@field children ht.right_click.MenuItemOptions[]
---@field desc string?
---@field callback fun()
---@field keys string[]
local MenuItemOptions = {}

---@class ht.right_click.MenuItem: ht.right_click.MenuItemOptions
---@field extra_keys string[]
---@field rendering_context table<string, any>
local MenuItem = {}

local disable_builtin_keys = {
  ["j"] = false,
  ["k"] = false,
  ["h"] = false,
  ["l"] = false,
  ["q"] = false,
  ["<ESC>"] = false,
  ["<C-c>"] = false,
  ["<CR>"] = false,
  ["<TAB>"] = false,
  ["<S-TAB>"] = false,
  ["<C-n>"] = false,
  ["<C-p>"] = false,
  ["<UP>"] = false,
  ["<DOWN>"] = false,
}

---@param opts ht.right_click.MenuItemOptions
---@return ht.right_click.MenuItem?
function MenuItem.new(opts)
  if opts[1] == nil then
    return nil
  end

  local text = MenuText.new(opts[1])
  local keys = {}
  local extra_keys = {}

  -- setup keys and pos from parsed text
  for _, key in ipairs(text:keys()) do
    keys[key] = true
  end

  -- setup keys from opts
  if opts.keys ~= nil then
    if type(opts.keys) == "string" then
      keys[opts.keys] = true
      extra_keys[#extra_keys + 1] = opts.keys
    elseif vim.tbl_isarray(opts.keys) then
      for _, key in ipairs(opts.keys) do
        keys[key] = true
        extra_keys[#extra_keys + 1] = key
      end
    end
  end

  keys = vim.tbl_extend("force", keys, disable_builtin_keys)
  table.sort(extra_keys)

  local children = {}
  if opts.children and vim.tbl_isarray(opts.children) then
    for _, child in ipairs(opts.children) do
      local c = MenuItem.new(child)
      if c then
        children[#children + 1] = c
      end
    end
  end

  return setmetatable({
    text = text,
    children = children,
    desc = opts.desc,
    callback = opts.callback,
    keys = UtilsTable.valid_keys(keys),
    extra_keys = extra_keys,
    rendering_context = {},
  }, { __index = MenuItem })
end

---@return number
function MenuItem:length()
  if self.text:is_separator() then
    return 0
  end

  local len = self.text:length()

  if #self.extra_keys > 0 then
    len = len + #self.extra_keys + 3
    for _, key in ipairs(self.extra_keys) do
      len = len + #key
    end
  end

  if #self.children > 0 then
    len = len + 2
  end

  return len
end

---@param text_width number
---@return NuiTreeNode
function MenuItem:as_nui_node(text_width)
  local nui_menu = require("nui.menu")
  if self.text:is_separator() then
    return nui_menu.separator(nil, { "-" })
  end

  local text = self.text:as_nui_line()
  local current_length = self.text:length()

  if #self.extra_keys > 0 then
    local hint = table.concat(self.extra_keys, "|")
    local space_count = text_width - 2 - current_length - #hint
    assert(space_count >= 0, "space_count should be positive")
    if space_count > 0 then
      text:append(string.rep(" ", space_count))
    end
    text:append(hint, "@variable.builtin")
    current_length = current_length + space_count + #hint
  end

  if #self.children > 0 then
    local space_count = text_width - current_length - 2
    assert(space_count >= 0, "space_count should be positive")
    if space_count > 0 then
      text:append(string.rep(" ", space_count))
    end
    text:append(" ▶")
  end

  return nui_menu.item(text, {
    menu_item = self,
  })
end

return MenuItem

File: src/core/components/menu-item.ts
import { uniqueArray } from "@core/utils/array";
import { MenuText } from "./menu-text";
import { NuiMenuMod } from "@extra/nui";
import { ContextMenu, ContextMenuContext } from "./context-menu";

const builtinKeys = [
  "j",
  "k",
  "h",
  "l",
  "q",
  "<ESC>",
  "<C-c>",
  "<CR>",
  "<TAB>",
  "<S-TAB>",
  "<C-n>",
  "<C-p>",
  "<UP>",
  "<DOWN>",
];

function removeBuiltinKeys(keys: string[]) {
  return keys.filter((key) => !builtinKeys.includes(key));
}

/**
 * A menu item component.
 */
export class MenuItem {
  public text: MenuText;
  public children: MenuItem[];
  public description?: string;
  public callback: (this: void) => void;
  public parent?: NuiMenu<MenuItemContext, ContextMenuContext>;

  private _keys: string[];

  constructor(
    text: string | MenuText,
    callback: () => void,
    options?: {
      children?: MenuItem[];
      description?: string;
      keys?: string[];
    }
  ) {
    this.text = typeof text === "string" ? new MenuText(text) : text;
    this.children = options?.children ?? [];
    this.description = options?.description;
    this.callback = callback;

    this._keys = removeBuiltinKeys(
      uniqueArray([...this.text.keys, ...(options?.keys ?? [])])
    );
  }

  /**
   * Get the menu item's keys.
   *
   * @returns The menu item's keys.
   */
  public get keys(): string[] {
    return this._keys;
  }

  public get textLength(): number {
    if (this.text.isSeparator()) {
      return 0;
    }
    return this.text.length;
  }

  public getPartLength() {
    if (this.text.isSeparator()) {
      return {
        textLength: 0,
        keysLength: 0,
        childrenLength: 0,
      };
    }

    const textLength = this.text.length + 3;

    let keysLength = 0;
    if (this.keys.length > 0) {
      keysLength = this.keys.length;
      for (const key of this.keys) {
        keysLength += key.length;
      }
    }

    let childrenLength = 0;
    if (this.children.length > 0) {
      childrenLength = 2; // ▸ mark
    }
    return {
      textLength,
      keysLength,
      childrenLength,
    };
  }

  asNuiTreeNode(opts: {
    textLength: number;
    keysLength: number;
    childrenLength: number;
  }): NuiTreeNode<MenuItemContext> {
    if (this.text.isSeparator()) {
      return NuiMenuMod.separator(undefined, {
        char: "-",
      });
    }

    let line = this.text.asNuiLine();

    fillSpaces(line, this.text.length, opts.textLength);

    if (this.keys.length > 0) {
      let hint = this.keys.join("|");
      line.append(hint, "@variable.builtin");
      fillSpaces(line, hint.length, opts.keysLength);
    }

    if (this.children.length > 0) {
      line.append("▸ ", "@variable.builtin");
    } else {
      fillSpaces(line, 0, opts.childrenLength);
    }

    return NuiMenuMod.item(line, {
      menuItem: this,
      parent: null,
    });
  }
}

function fillSpaces(line: NuiLine, length: number, expect: number) {
  if (expect > length) {
    line.append(" ".repeat(expect - length));
  }
}

export type MenuItemContext = {
  menuItem: MenuItem;
  parent: ContextMenu | null;
};
